src/GraphQL/Validation/Rules/NoFragmentCycles.cs

```diff
@@ -13,33 +13,41 @@ namespace GraphQL.Validation.Rules
     /// </summary>
     public class NoFragmentCycles : IValidationRule
     {
-        /// <summary>
-        /// Returns a static instance of this validation rule.
-        /// </summary>
-        public static readonly NoFragmentCycles Instance = new NoFragmentCycles();
-
-        /// <inheritdoc/>
-        /// <exception cref="NoFragmentCyclesError"/>
-        public Task<INodeVisitor> ValidateAsync(ValidationContext context)
+        private sealed class NoFragmentCyclesData
         {
-            // Tracks already visited fragments to maintain O(N) and to ensure that cycles
-            // are not redundantly reported.
-            var visitedFrags = new LightweightCache<string, bool>(key => false);
+            // Tracks already visited fragments to maintain O(N) and to ensure that cycles are not redundantly reported.
+            public LightweightCache<string, bool> VisitedFrags { get; set; } = new LightweightCache<string, bool>(key => false);
 
             // Array of AST nodes used to produce meaningful errors
-            var spreadPath = new Stack<FragmentSpread>();
+            public Stack<FragmentSpread> SpreadPath = new Stack<FragmentSpread>();
 
             // Position in the spread path
-            var spreadPathIndexByName = new LightweightCache<string, int>(key => -1);
+            public LightweightCache<string, int> SpreadPathIndexByName = new LightweightCache<string, int>(key => -1);
+        }
+
+        /// <summary>
+        /// Returns a static instance of this validation rule.
+        /// </summary>
+        public static readonly NoFragmentCycles Instance = new NoFragmentCycles();
 
-            return new MatchingNodeVisitor<FragmentDefinition>(node =>
+        private static readonly Task<INodeVisitor> _task = new EnterLeaveListener(_ =>
+            {
+                _.Match<Document>((_, context) => context.Set<NoFragmentCycles>(new NoFragmentCyclesData()));
+                _.Match<FragmentDefinition>((node, context) =>
                 {
-                    if (!visitedFrags[node.Name])
+                    var data = context.Get<NoFragmentCycles, NoFragmentCyclesData>();
+                    if (!data.VisitedFrags[node.Name])
                     {
-                        detectCycleRecursive(node, spreadPath, visitedFrags, spreadPathIndexByName, context);
+                        detectCycleRecursive(node, data.SpreadPath, data.VisitedFrags, data.SpreadPathIndexByName, context);
                     }
-                }).ToTask();
-        }
+                });
+            },
+            shouldRun: context => context.Document.Fragments.Count > 0
```

Comment:
Nice 👍

Reply:
It's a pity that not all the rules are so simple to bypass. Although the possibility of such a check already shows us how expensive the current implementation of the rules is.